### ISSUE TITLE
Addon Request: Reddit (expansion of existing RSS support)

### ISSUE THREAD

**Issue by shiribailem** [friendica/friendica#13063]
### Is the feature request related to a problem? Please describe.

Reddit has an API and has also started blocking bots for RSS (ie. Friendica connections), it's a prime candidate for a reader/connector addon.

### Describe the feature you'd like

A Reddit addon would allow easier adding of reddit groups and users (being able to potentially search username@reddit.com or r/forum@reddit.com). Additionally with an addon it can do things like sync likes/dislikes to upvotes/downvotes and can have a user setting to choose between only showing posts over a certain score or all posts.

On top of that, if they connect to a reddit account it would allow them to upvote/downvote via like/dislike and comment on posts. (Making whole new posts is a much bigger feature ask of course)

### Describe alternatives you've considered

The rss feeds exist, but they involve knowing the trick and a manual URL edit. Also because Reddit is trying to crack down on scraping, especially with the concerns around AI training data, they're cracking down on bot access to RSS feeds. I've found that it's inconsistent at best adding Reddit RSS feeds because of this now.

Additionally, this allows little options for filtering, sorting noise as well as no ability to directly reply or follow conversations on a post.

### Additional context

**Comment by MrPetovan** [comment]
Unrelated to the feature request: What other RSS access exists besides by bots? It isn't exactly human-readable.

**Comment by shiribailem** [comment]
Bots being automated pull without a human element vs someone manually opening and loading an RSS reader. (HTML isn't exactly human readable either?)

They flag based on user agents and usage rates, coming from a server (especially if it's a busy one) will flag hard. They then impose a scaling rate limit (more requests, longer waits) in seconds.

**Comment by annando** [comment]
Reddit considers charging for the API access: https://techcrunch.com/2023/04/18/reddit-will-begin-charging-for-access-to-its-api/

As long as they haven't clarified their API rules, there is the risk that any work will be in vain.

**Comment by shiribailem** [comment]
I saw that, *but* reading through their statement on it I'm pretty sure they'd consider Friendica a fair case. The big angle they're talking about charging for is for the purpose of data processing (ie. AI training data), and they mostly talk about having issue with things that take but don't give (ie. read but don't post)

But also fair that they haven't made a clear official statement of a free version. I mostly made this because I saw there wasn't one and it felt apt to at least start the discussion.

Based on their statements I expect it to be a free rate limited tier with a rate that'll be good for most purposes and probably won't be an issue outside of huge servers or the most popular of bots.

**Comment by bkil** [comment]
I recommend migrating your communities to open platforms such as Friendica, Hubzilla or Lemmy as an alternative instead.

**Comment by shiribailem** [comment]
@bkil that's not an answer, it's more just an argument against having any connector add-ons at all.

**Comment by bkil** [comment]
We can only guess until the wording within the terms of service is updated, but based on **my** reading of their statements, usage of the API for exfiltration (such as allowing Reddit users to escape their walled garden to leave for Friendica but still monitor what's happening) would be specifically prohibited.

> Mr. Huffman said Reddit’s A.P.I. would still be free to developers who wanted to build applications that helped people use Reddit.

In **my** reading, this only allows you to build a mobile application to search for posts on Reddit, to display posts on Reddit or submit posts to Reddit (i.e., implementing a dedicated Reddit client that allows users of Reddit to use Reddit).

**Comment by MrPetovan** [comment]
Post connectors still necessitate an account on the third-party platform so arguably a Friendica connector is helping Reddit users use Reddit.

**Comment by shiribailem** [comment]
Just going to put this as a through line: unless you're actively working on this, I don't think anybody should be making any more comments on this request until they have released their terms.

This is to avoid spamming notifications and pushing this to the top for idle discussion.

**Comment by annando** [comment]
@tmajibon since I'm not using Reddit, it would be great if you could have got the API on your watchlist. It would be great if you then could add some notes here. Then I will have got a look at what the API is capable of and what can be done.

**Comment by shiribailem** [comment]
Funny thing, I looked and silly me they already updated their API terms a couple of weeks ago with what they discussed:

https://www.redditinc.com/policies/developer-terms - Updated 2023-04-18

> access or use the Reddit Services and Data through any means (including by accessing our API or indexing, caching, or crawling our Reddit Services and Data) to train large language, artificial intelligence, or other algorithmic models or related services without our permission; or

and https://www.redditinc.com/policies/data-api-terms - updated 2023-04-18 as well

**Comment by shiribailem** [comment]
API documentation is available here: https://www.reddit.com/dev/api/

Most of the routines in there would probably not be used due to lack of feature parity. Upvote and Downvote would be mapped to like and dislike, mapping subreddits as forums and users as regular users. Private messages map as well, but no general post federation (for general posting, it *might* be valid to post to a tagged subreddit and direct message an explicitly tagged user, but not PM them if it's in a comment).

Because of the nature of how reddit operates vs Friendica, I mentioned before there should probably be a user filter on posts as Reddit is designed for users to post anything/everything and have voting move it to the top. Some people view based on newest posted, but most view off of "hot" based on the algorithm. There are different calls in the API to select between these schemes and they should probably be selectable by the user, but if not feasible they should default to "hot" or an admin setting.

There is also a call for editing that can be mapped to update posts if they're edited in friendica.

And there's a follow feature that can be used potentially to improve performance and streamline requests, perhaps by allowing the admin to have a bot account for read-only access that follows all subreddits/users/etc that server users are following.

**Comment by bkil** [comment]
How do you plan to handle Google reCAPTCHA? I have always been greeted with that when I tried to post from the web site of Reddit.

**Comment by shiribailem** [comment]
@bkil there's at least one call in there specifically for identifying what does and doesn't need it at any point in time ( https://www.reddit.com/dev/api/#GET_api_needs_captcha ). Given that bots operate using this API I'm certain there's enough flexibility to work around this.

I did a quick search and from what I can see briefly around the captcha requirements are based on *new* bot accounts until they receive enough karma (recognition by other users for positive contributes). As a last resort, it could be maybe treated as a message to the relevant user and/or admin to put in the captcha and deferring the relevant API or APIs until then.

But as I understand it, this means it'll probably not be affecting read-only calls, and will only affect posting calls when they're using it with a *new* account.

**Comment by bkil** [comment]
I'm sure there is some complicated dynamic risk assessment built in as my Reddit account had a positive score and it's quite a few years old. It kept doing it despite my upvotes going up after each post. I probably didn't click on enough advertisements or something.

Could you perhaps share the CAPTCHA requirements that you had found? I would be curious to see their algorithms.

**Comment by shiribailem** [comment]
@bkil I didn't see any specifics and they were >5 years old, nothing I saw was official.

Regardless, dealing with Captcha's is going to be a required part of handling write APIs. And I suspect it's per-account, which makes it easier.

It might be worth a potential core feature system for Captchas because there's a lot of places they can crop up, potentially on RSS pages or just about any service behind Cloudflare. (As mentioned before, I think it could be done with a direct message to the relevant users, delete the message if it gets resolved by someone else or simply by time, and the message could have a link to a Captcha handling page or possibly include one as an embed?)

**Comment by shiribailem** [comment]
Just dropping this here, this is probably completely done:

https://www.reddit.com/r/apolloapp/comments/12ram0f/had_a_few_calls_with_reddit_today_about_the/

**Comment by annando** [comment]
Thanks for the link. Sadly this ` tl;dr: Paid API coming.` is the blocker here. It doesn't seem as if they will allow some reasonable access for free. At least I will not invest time into it then. (Which is sad, because I like these connectors)

**Comment by shiribailem** [comment]
Yeah, this contradicts all their messaging about this change... including "reasonable costs"

**Comment by annando** [comment]
See also here: https://www.reddit.com/r/apolloapp/comments/13ws4w3/had_a_call_with_reddit_to_discuss_pricing_bad/
